Stratis.Bitcoin/Notifications/TransactionReceiver.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System.Linq;
+using System.Threading.Tasks;
+using NBitcoin;
+using NBitcoin.Protocol;
+using NBitcoin.Protocol.Behaviors;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Notifications
+{
+    /// <summary>
+    /// This class receives transaction messages from other nodes.
+    /// </summary>
+    public class TransactionReceiver : NodeBehavior
+    {
+        private readonly TransactionNotification transactionNotification;
+        private readonly TransactionNotificationProgress notifiedTransactions;
+
+        public TransactionReceiver(TransactionNotification transactionNotification, TransactionNotificationProgress notifiedTransactions)
+        {
+            this.transactionNotification = transactionNotification;
+            this.notifiedTransactions = notifiedTransactions;
+        }
+
+        protected override void AttachCore()
+        {
+            this.AttachedNode.MessageReceived += AttachedNode_MessageReceived;
+        }
+
+        protected override void DetachCore()
+        {
+            this.AttachedNode.MessageReceived -= AttachedNode_MessageReceived;
+        }
+
+        private async void AttachedNode_MessageReceived(Node node, IncomingMessage message)
+        {
+            Guard.Assert(node == this.AttachedNode); // just in case
+
+            // check the type of message received.
+            // we're only interested in Inventory and Transaction messages.
+            var invPayload = message.Message.Payload as InvPayload;
+            if (invPayload != null)
+            {
+                await this.ProcessInvAsync(node, invPayload);
+                return;
+            }
+
+            var txPayload = message.Message.Payload as TxPayload;
+            if (txPayload != null)
+            {
+                this.ProcessTxPayload(txPayload);                
+            }            
+        }
+
+        private void ProcessTxPayload(TxPayload txPayload)
+        {
+            var transaction = txPayload.Object;
+            var trxHash = transaction.GetHash();
+
+            if (this.notifiedTransactions.TransactionsReceived.ContainsKey(trxHash))
+            {
+                return;
+            }
+
+            // send the transaction to the notifier
+            transactionNotification.Notify(transaction);
+            this.notifiedTransactions.TransactionsReceived.TryAdd(trxHash, trxHash);
+        }
+
+        private async Task ProcessInvAsync(Node node, InvPayload invPayload)
+        {            
+            var txs = invPayload.Inventory.Where(inv => inv.Type.HasFlag(InventoryType.MSG_TX));
+            
+            // get the transactions in this inventory that have never been received - either because new or requested.
+            var newTxs = txs.Where(t => this.notifiedTransactions.TransactionsReceived.All(ts => ts.Key != t.Hash)).ToList();
```

Comment:
Can't really use contains() here as what I'm trying to get is a list of received transactions that don't appear in the list of transactions already received, not a single element.